llvm-build/utils.ps1

```diff
@@ -0,0 +1,278 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+if (Test-Path function:\exec) {
+    # already included
+    return
+}
+
+if (!(Test-Path function:\Get-RepoRoot)) {
+    # git revparse uses cwd. E2E builds use a different
+    # working dir, so we pin it to out repo (submodule in E2E)
+    function Get-RepoRoot {
+        exec -wd $PSScriptRoot {
+            git rev-parse --show-toplevel
+        }
+    }
+}
+
+# Fix temp path for non-windows platforms if missing
+if (!(Test-Path env:\TEMP)) {
+    $env:TEMP = [System.IO.Path]::GetTempPath()
+}
+
+# Test whether we build the compiler components.
+function Test-BuildCompiler {
+    # By default we want to build the compiler
+    # If no env var is defined we default to true
+    if (!(Test-Path env:\ENABLE_COMPILER)) {
+        $true
+    }
+    else {
+        # The env var exists, we use its value
+        # Coerce falsy values and negate to keep the defalt $true
+        $env:ENABLE_COMPILER -ne $false
+    }
+}
+
+# Test whether we build the VSIX components.
+function Test-BuildVsix {
+    ($Env:ENABLE_VSIX -ne "false") -and $IsWindows
+}
+
+# Test whether we build the LLVM components.
+function Test-BuildLlvmComponents {
+    ((Test-Path env:\ENABLE_LLVM_BUILDS) -and ($env:ENABLE_LLVM_BUILDS -eq $true))
+}
+
+####
+# Utilities
+####
+
+# executes the sciptblock. If the working directory is specitied, the block is
+# exeucted in that directory. If the $LASTEXITCODE from the block is not 0,
+# an exception is thrown.
+function exec {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        [scriptblock]$cmd,
+        [string]$errorMessage = "Failed to execute command",
+        [Alias("wd")]
+        [string]$workingDirectory = $null
+    )
+    try {
+
+        if ($workingDirectory) {
+            Push-Location -Path $workingDirectory
+        }
+
+        $global:lastexitcode = 0
+        & $cmd
+        if ($global:lastexitcode -ne 0) {
+            throw "exec: $errorMessage"
+        }
+    }
+    finally {
+        if ($workingDirectory) {
+            Pop-Location
+        }
+    }
+}
+
+# tests whether a condition is true. Throws an exception with the specified error message
+# if the condition check fails.
+function Assert {
+    [CmdletBinding()]
+    param(
+        [Parameter(Mandatory = $true)]
+        $conditionToCheck,
+
+        [Parameter(Mandatory = $true)]
+        [string]$failureMessage
+    )
+
+    if (-not $conditionToCheck) {
+        throw ('Assert: {0}' -f $failureMessage)
+    }
+}
+
+# returns true if the script is running on a build agent, false otherwise
+function Test-CI {
+    if (Test-Path env:\TF_BUILD) {
+        $true
+    }
+    elseif ((Test-Path env:\CI)) {
+        $env:CI -eq $true
+    }
+    else {
+        $false
+    }
+}
+
+# Writes an Azure DevOps message with default debug severity
+function Write-Vso {
```

Comment:
Why `Vso`? What is the relation between the Azure DevOps and the `Vso`?

Reply:
Actually, until you called attention to it I hadn't even noticed it because some part of me is used to the old name... this terminology is a holdover from the days when Azure DevOps was called "Visual Studio Online." When it was first announced, git-compatible services via Azure was named [Visual Studio Online (VSO)](https://docs.microsoft.com/en-us/archive/msdn-magazine/2014/january/visual-studio-2013-introducing-visual-studio-online), then later rebranded to Visual Studio Team Services (VSTS), and finally in 2018 renamed to Azure DevOps. Interestingly, you can still see this legacy in the [default format for logging commands](https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#logging-command-format):
`##vso[area.action property1=value;property2=value;...]message`
However, DevOps added an alternate logging format that doesn't use the `vso` anymore, and that happens to be what this utility uses:
`##[command]message`.

Still, I'm not opposed to renaming it for clarity since not everyone comes to DevOps with that naming baggage.